io_buffer.c

```diff
@@ -0,0 +1,810 @@
+{
+#if defined(_WIN32)
+    void * base = VirtualAlloc(0, size, MEM_COMMIT, PAGE_READWRITE);
+
+    if (!base) {
+        rb_sys_fail("IO_Buffer_map:VirtualAlloc");
+    }
+#else
+    void * base = mmap(NULL, size, PROT_READ | PROT_WRITE, MAP_ANON | MAP_PRIVATE, -1, 0);
+
+    if (base == MAP_FAILED) {
+        rb_sys_fail("IO_Buffer_map:mmap");
+    }
+#endif
+
+  return base;
+}
+
+static inline void IO_Buffer_unmap(void* base, size_t size)
+{
+#ifdef _WIN32
+    VirtualFree(base, 0, MEM_RELEASE);
+#else
+    munmap(base, size);
+#endif
+}
+
+static void IO_Buffer_initialize(struct rb_IO_Buffer *data, void *base, size_t size, enum rb_IO_Buffer_Flags flags, VALUE source)
+{
+    data->flags = flags;
+    data->size = size;
+
+    if (base) {
+        data->base = base;
+    } else {
+        if (data->flags & RB_IO_BUFFER_INTERNAL) {
+            data->base = malloc(data->size);
+        } else if (data->flags & RB_IO_BUFFER_MAPPED) {
+            data->base = IO_Buffer_map(data->size);
+        }
+    }
+
+    if (!data->base) {
+        rb_raise(rb_eRuntimeError, "Could not allocate buffer!");
+    }
+
+    data->source = source;
+}
+
+static int IO_Buffer_clear(struct rb_IO_Buffer *data)
+{
+    if (data->base) {
+        if (data->flags & RB_IO_BUFFER_INTERNAL) {
+            free(data->base);
+        }
+
+        if (data->flags & RB_IO_BUFFER_MAPPED) {
+            IO_Buffer_unmap(data->base, data->size);
+        }
+
+         data->base = NULL;
+
+         return 1;
+    }
+
+    return 0;
+}
+
+void rb_IO_Buffer_Type_mark(void *_data)
+{
+    struct rb_IO_Buffer *data = _data;
+    rb_gc_mark(data->source);
+}
+
+void rb_IO_Buffer_Type_free(void *_data)
+{
+    struct rb_IO_Buffer *data = _data;
+
+    IO_Buffer_clear(data);
+
+    free(data);
+}
+
+size_t rb_IO_Buffer_Type_size(const void *_data)
+{
+    const struct rb_IO_Buffer *data = _data;
+    size_t total = sizeof(struct rb_IO_Buffer);
+
+    if (data->flags) {
+        total += data->size;
+    }
+
+    return total;
+}
+
+static const rb_data_type_t rb_IO_Buffer_Type = {
+    .wrap_struct_name = "IO::Buffer",
+    .function = {
+        .dmark = rb_IO_Buffer_Type_mark,
+        .dfree = rb_IO_Buffer_Type_free,
+        .dsize = rb_IO_Buffer_Type_size,
+    },
+    .data = NULL,
+    .flags = RUBY_TYPED_FREE_IMMEDIATELY,
+};
+
+VALUE rb_IO_Buffer_Type_allocate(VALUE self)
+{
+    struct rb_IO_Buffer *data = NULL;
+    VALUE instance = TypedData_Make_Struct(self, struct rb_IO_Buffer, &rb_IO_Buffer_Type, data);
+
+    data->base = NULL;
+    data->size = 0;
+    data->flags = 0;
+    data->source = Qnil;
+
+    return instance;
+}
+
+VALUE rb_IO_Buffer_Type_for(VALUE klass, VALUE string)
+{
+  VALUE instance = rb_IO_Buffer_Type_allocate(klass);
+
+  struct rb_IO_Buffer *data = NULL;
+  TypedData_Get_Struct(instance, struct rb_IO_Buffer, &rb_IO_Buffer_Type, data);
+
+  IO_Buffer_initialize(data, RSTRING_PTR(string), RSTRING_LEN(string), 0, string);
```

Comment:
This seems too dangerous, as if the string is later appended/shrinked/resized then its RSTRING_PTR() may change, and you would get segfaults/memory corruption if the contents of the IO::Buffer (which the old RSTRING_PTR, which has been free'd) is accessed then.
To fix I think we'd need either a way to prevent the String to be resized (but might be unintuitive for the Ruby side of things) or copy the contents of that String.

Or maybe on every IO::Buffer method & access set the `base`/`size` from RSTRING_PTR/RSTRING_LEN.
That could still be problematic if another thread would concurrently modify the String, but that's a general and unavoidable problem of RSTRING_PTR().

Reply:
We validate this before every usage, check `rb_IO_Buffer_get_*`. That's the best we can do. Otherwise, we can remove it. I'm on the fence, but I think it's useful enough to keep it, but we can see if it actually ends up being a useful interface or remove it.